sdk/mobile-center-updates/src/main/java/com/microsoft/azure/mobile/updates/Updates.java

```diff
@@ -0,0 +1,971 @@
+             *
+             * We still want to generate the notification: if we can find the data in preferences
+             * that means they were not deleted, and thus that the sdk was not disabled.
+             */
+            MobileCenterLog.debug(LOG_TAG, "Process download completion id=" + mDownloadId);
+            if (Updates.this.mContext == null) {
+                MobileCenterLog.debug(LOG_TAG, "Called before onStart, init storage");
+                StorageHelper.initialize(mContext);
+            }
+
+            /* Check intent data is what we expected. */
+            long expectedDownloadId = StorageHelper.PreferencesStorage.getLong(PREFERENCE_KEY_DOWNLOAD_ID);
+            if (expectedDownloadId > 0 && expectedDownloadId != mDownloadId) {
+                MobileCenterLog.warn(LOG_TAG, "Ignoring completion for a download we didn't expect, id=" + mDownloadId);
+                return null;
+            }
+
+            /* Check if download successful. */
+            DownloadManager downloadManager = (DownloadManager) mContext.getSystemService(DOWNLOAD_SERVICE);
+            Uri uriForDownloadedFile = downloadManager.getUriForDownloadedFile(mDownloadId);
+            if (uriForDownloadedFile != null) {
+
+                /* Build install intent. */
+                MobileCenterLog.debug(LOG_TAG, "Download was successful for id=" + mDownloadId + " uri=" + uriForDownloadedFile);
+                Intent intent = getInstallIntent(uriForDownloadedFile);
+                boolean installerFound = false;
+                if (intent.resolveActivity(mContext.getPackageManager()) == null) {
+                    if (Build.VERSION.SDK_INT < Build.VERSION_CODES.N) {
+                        Cursor cursor = downloadManager.query(new DownloadManager.Query().setFilterById(mDownloadId));
+                        if (cursor != null) {
+                            if (cursor.moveToNext()) {
+                                uriForDownloadedFile = getFileUriOnOldDevices(cursor);
+                                intent = getInstallIntent(uriForDownloadedFile);
+                                installerFound = intent.resolveActivity(mContext.getPackageManager()) != null;
+                            }
+                            cursor.close();
+                        }
+                    }
+                } else {
+                    installerFound = true;
+                }
+                if (!installerFound) {
+                    MobileCenterLog.error(LOG_TAG, "Installer not found");
+                    completeWorkflow(this);
+                    return null;
+                }
+
+                /* Exit check point. */
+                Activity activity;
+                try {
+                    activity = getForegroundActivityWithStateCheck(this);
+                } catch (IllegalStateException e) {
+
+                    /* If we were canceled, exit now. */
+                    return null;
```

Comment:
Can you add a comment for that? *If we were cancelled and cleaned up, exit now.*